src/irr/asset/CCPUMeshPacker.h

```diff
@@ -46,85 +67,243 @@ PackedMeshBufferData CCPUMeshPacker<MDIStructType>::alloc(const ICPUMeshBuffer c
 		{
 			_IRR_DEBUG_BREAK_IF(true);
-			return invalidPackedMeshBufferData;
+			return invalidReservedAllocationMeshBuffers;
 		}
 	}
 
 	//validation
-	const auto& mbVtxInputParams = meshBuffer->getPipeline()->getVertexInputParams();
-	for (uint16_t attrBit = 0x0001, location = 0; location < 16; attrBit <<= 1, location++)
+	for (auto it = begin; it != end; it++)
 	{
-		if (!(attrBit & mbVtxInputParams.enabledAttribFlags))
-			continue;
+		const auto& mbVtxInputParams = (*it)->getPipeline()->getVertexInputParams();
+		for (uint16_t attrBit = 0x0001, location = 0; location < SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT; attrBit <<= 1, location++)
+		{
+			if (!(attrBit & mbVtxInputParams.enabledAttribFlags))
+				continue;
 
-		//assert((attrBit & m_outVtxInputParams.enabledAttribFlags));
+			//assert((attrBit & m_outVtxInputParams.enabledAttribFlags));
 
-		if (mbVtxInputParams.attributes[location].format != m_outVtxInputParams.attributes[location].format ||
-			mbVtxInputParams.bindings[mbVtxInputParams.attributes[location].binding].inputRate != m_outVtxInputParams.bindings[location].inputRate)
-		{
-			_IRR_DEBUG_BREAK_IF(true);
-			return invalidPackedMeshBufferData;
+			if (mbVtxInputParams.attributes[location].format != m_outVtxInputParams.attributes[location].format ||
+				mbVtxInputParams.bindings[mbVtxInputParams.attributes[location].binding].inputRate != m_outVtxInputParams.bindings[location].inputRate)
+			{
+				_IRR_DEBUG_BREAK_IF(true);
+				return invalidReservedAllocationMeshBuffers;
+			}
 		}
 	}
+	
+	const size_t idxCnt = std::accumulate(begin, end, 0ull, [](size_t init, ICPUMeshBuffer* mb) { return init + mb->getIndexCount(); });
+	const size_t vtxCnt = std::accumulate(begin, end, 0ull, [](size_t init, ICPUMeshBuffer* mb) { return init + mb->calcVertexCount(); });
+
+	const uint32_t minIdxCntPerPatch = m_minTriangleCountPerMDIData * 3;
+	const uint32_t possibleMDIStructsNeededCnt = core::roundUp<uint32_t>(idxCnt, minIdxCntPerPatch);
+
+	uint32_t MDIAllocAddr       = INVALID_ADDRESS;
+	uint32_t idxAllocAddr       = INVALID_ADDRESS;
+	uint32_t vtxAllocAddr       = INVALID_ADDRESS;
+	uint32_t perInsVtxAllocAddr = INVALID_ADDRESS;
+
+	//actually it will not work at all if m_MDIDataAlctrResSpc == nullptr or m_idxBuffAlctrResSpc == nullptr.. TODO
+	if (m_MDIDataAlctrResSpc)
+	{
+		MDIAllocAddr = m_MDIDataAlctr.alloc_addr(possibleMDIStructsNeededCnt, 1u);
+		if (MDIAllocAddr == INVALID_ADDRESS)
+			return invalidReservedAllocationMeshBuffers;
+	}
+	
+	if (m_idxBuffAlctrResSpc)
+	{
+		idxAllocAddr = m_idxBuffAlctr.alloc_addr(idxCnt, 1u);
+		if (idxAllocAddr == INVALID_ADDRESS)
+			return invalidReservedAllocationMeshBuffers;
+	}
+	
+	if (m_vtxBuffAlctrResSpc)
+	{
+		vtxAllocAddr = m_vtxBuffAlctr.alloc_addr(vtxCnt, 1u);
+		if (vtxAllocAddr == INVALID_ADDRESS)
+			return invalidReservedAllocationMeshBuffers;
+	}
+	
+	if (m_perInsVtxBuffAlctrResSpc)
+	{
+		perInsVtxAllocAddr = m_perInsVtxBuffAlctr.alloc_addr(vtxCnt, 1u);
```

Comment:
Nope, you'd need to iterate through the meshbuffers and accumulate `getInstanceCount`

Reply:
done?

Reply:
?